### ISSUE TITLE
[Epic] Migrate to Webpack

### ISSUE THREAD

**Issue by alexgibson** [mozilla/bedrock#8985]
## Description

This is an epic to help break down the work that would be required to migrate bedrock to Webpack.

**Comment by alexgibson** [comment]
Dependencies:

- Research switching to Webpack (https://github.com/mozilla/bedrock/issues/8956)
- Run CSS/JS linting and tests via npm scripts (https://github.com/mozilla/bedrock/issues/8479)
- Ship an MVP Webpack config (https://github.com/mozilla/bedrock/issues/8986)
- Add a webpack build option for compiling ES2015+ using Babel (https://github.com/mozilla/bedrock/issues/8987)

**Comment by alexgibson** [comment]
I've given a lot of thought to using Webpack for bedrock, and at this time I'm not convinced that it's the best option for a front-end build tool given the nature of www.mozilla.org. It would be a considerable amount of effort to switch at this time, and whilst it would solve some problems it is not free of creating new problems itself.

The main reasons why I don't think it's worth switching right now:

- We have a vast number of pages each with unique CSS/JS bundles. It is certainly possible to suport multiple page sites with Webpack, but it is optimized mainly for single page apps or websites that share common assets on all pages. As a result, the number of files it needs to rebuild at a high frequency whilst in dev mode is not really very efficient.
- Because we rely on Django for hashing assets via collectstatic, we can't easily make proper use of advanced features such as code splitting.
- Given the number of static assets we serve in bedrock, Webpack still requires high memory and CPU usage whilst in dev mode. This is much the same as the situation we have with Gulp, and does not really improve things (as far as my testing has shown). I had to split out static assets into a separate config that aren't watched, else docker sometimes throws an out of memory error.
- We still need to support bundling for some browsers older than Webpack supports. We can get around this using plugins, but it means we lose out on some of the other advantages webpack offers by creating the need to write more custom config than would otherwise be necessary.

Given the above, I think it may be a more effective use of time if we concentrate (for now at least) on simplifying our existing Gulp configuration. In https://github.com/mozilla/bedrock/issues/8479 we moved all our front-end linting and testing to NPM scripts. Doing the same for static assets, and leaving Gulp only for CSS/JS files might vastly reduce the number of files the process is watching. I'm sure there are other improvements / simplifications we could make too.
